File: frontend/src/objects/contest/request/CreateContestRequest.ts
import type { ContestDescription } from '@/objects/contest/ContestDescription'
import type { ContestSlug } from '@/objects/contest/ContestSlug'
import type { ContestTitle } from '@/objects/contest/ContestTitle'
import type { ResourceAccessPolicy } from '@/objects/shared/access/ResourceAccessPolicy'

/** 创建比赛请求体；时间为后端约定的字符串格式，权限策略显式传入。 */
export type CreateContestRequest = {
  slug: ContestSlug
  title: ContestTitle
  description: ContestDescription
  startAt: string
  endAt: string
  accessPolicy: ResourceAccessPolicy
}

File: backend/src/main/scala/domains/contest/objects/request/CreateContestRequest.scala
package domains.contest.objects.request

import domains.contest.objects.*
import io.circe.{Decoder, Encoder}
import io.circe.generic.semiauto.{deriveDecoder, deriveEncoder}
import shared.objects.access.ResourceAccessPolicy

import java.time.Instant

/** 创建比赛请求体，包含基础信息、时间范围和访问策略。 */
final case class CreateContestRequest(
  slug: ContestSlug,
  title: ContestTitle,
  description: ContestDescription,
  startAt: Instant,
  endAt: Instant,
  accessPolicy: ResourceAccessPolicy
)

/** 提供创建比赛请求体 JSON codec。 */
object CreateContestRequest:
  given Encoder[CreateContestRequest] = deriveEncoder[CreateContestRequest]
  given Decoder[CreateContestRequest] = deriveDecoder[CreateContestRequest]
